### ISSUE TITLE
Support column specification for INSERT

### ISSUE THREAD

**Issue by JLDLaughlin** [MaterializeInc/materialize#3911]
Details from https://github.com/MaterializeInc/materialize/issues/3903.

We should probably support this use case:
```
insert into t (b, a) values (1, 2);
```

**Comment by benesch** [comment]
It occurs to me that we can also support 

```sql
create table t (a int, b int)
insert into t (a) values (1)
```

because the, er, default default is `NULL`. So the `insert` above inserts (1, null). Don't even need to special case if `b` is marked `NOT NULL` because it'll just fail the existing validation in the coordinator.

**Comment by elindsey** [comment]
So, I poked around at this today. Writing up a few notes here.

Most of this is around statement.rs' `handle_insert`. Downstream in coord.rs when we hand off to the backend we're relying on the table's column ordering (see `sequence_insert`), so it doesn't seem easy to rearrange the columns in the expr itself without having those changes ripple out. It'd probably be easiest to instead rearrange the parameters to match the table's column ordering. 

That looks mostly straightforward, but has a few annoying bits:
1. The best place to plug in this logic is `handle_insert`, but that means that anything not calling that function is going to have a bad time. Specifically, since slt calls describe then goes off an grabs the prepared statement/executes it itself, it will be buggy when doing these kinds of inserts (similar to the original bug report, it'll end up misbinding params to the wrong columns). Maybe that's fine if slt is the only thing calling into materialize in that weird way; it doesn't really look like it's a generally supported API.
2. The StatementContext holds a refcounted copy of type info for the parameters as originally ordered. Need to see if this should be rewritten to reflect the parameter reordering, and what kind of effects that might have when done in the middle of planning.

**Comment by benesch** [comment]
Yep, `handle_insert` is definitely the right place for this code to live. (I think actually in one of the functions it calls, but pretty close by.) A few quick thoughts follow. 

Not sure if you noticed, but SLT will actually route inserts to your local PostgreSQL instance. See: https://github.com/MaterializeInc/materialize/blob/078bbf6cdb1a87f5f4803c086651e8eb4b729003/src/coord/src/coord.rs#L819-L839

So actually for SLT inserts don't (presently) matter at all. The goal is to eventually get rid of symbiosis mode when we can handle the full battery of `CREATE TABLE`/`INSERT`/`UPDATE`/`DELETE`, but we're not there yet. (This issue helps though!)

There is a distinction between what SQL calls "parameters" and... what I think you are using the term "parameters" to mean. Specifically, in a statement like

```
INSERT INTO t VALUES ($2, 'a', 'b', $1)
```

only the `$1` and `$2` are called "parameters". The `'a'` and `'b'` are just... "values expressions" or something. Not sure there's a great term.

Anyway, given something like

```
INSERT INTO t (col2, col1, col3) VALUES (expr_for_2, expr_for_1, expr_for_3)
```

you'll want to take the planned `VALUES` expression and hit it with a project to permute the columns into place. In the example above I think it would look something like:

```rust
expr.project(vec![1, 0, 2])
```

So it should mostly be a matter of determine the correct "unpermutation" and slapping the expression with that projection in the right spot. Shouldn't need to reorder the `$1` parameters at all; those will just fall into place when you've permuted the expr!

**Comment by benesch** [comment]
Handling defaults like I mentioned in a much earlier comment is going to be trickier, so I'd recommend punting on that for now, and requiring that the column specification includes each column in the table exactly once.

**Comment by cuongdo** [comment]
@elindsey Is this resolved by #4354? If so, please close this issue.

Also, you can auto-close issues upon merge of a PR if the PR or commit description contains something like `Fixes #4354`. See the [GitHub docs](https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords) for more examples.

**Comment by elindsey** [comment]
All wrapped up except for supporting partial colspecs with default arguments - I'll file that into a separate task.
